File: packages/jetbrains/src/main/kotlin/com/hermes/ide/tools/Terminal.kt
package com.hermes.ide.tools

import com.hermes.ide.ToolResultPayload
import com.intellij.openapi.project.Project
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Execute terminal commands using the IDE's terminal or process execution.
 *
 * Uses ProcessBuilder for simple commands with output capture.
 * The agent never touches the filesystem — commands execute locally via the IDE.
 */
class TerminalOps(private val project: Project) {

    /**
     * Run a shell command and return its output.
     *
     * Uses ProcessBuilder with configurable working directory.
     * For long-running commands, uses a timeout of 2 minutes.
     */
    fun runCommand(command: String, cwd: String?): ToolResultPayload {
        return try {
            val workingDir = if (cwd != null) {
                File(cwd)
            } else {
                project.basePath?.let { File(it) } ?: File(System.getProperty("user.dir"))
            }

            val process = ProcessBuilder()
                .command("sh", "-c", command)
                .directory(workingDir)
                .redirectErrorStream(true)
                .start()

            val completed = process.waitFor(120, TimeUnit.SECONDS)
            val output = process.inputStream.bufferedReader().readText()

            if (!completed) {
                process.destroyForcibly()
                ToolResultPayload(
                    ok = false,
                    content = output,
                    error = "Command timed out after 120s",
                )
            } else if (process.exitValue() != 0) {
                ToolResultPayload(
                    ok = false,
                    content = output,
                    error = "Command exited with code ${process.exitValue()}",
                )
            } else {
                ToolResultPayload(
                    ok = true,
                    content = output.ifEmpty { "(command completed with no output)" },
                )
            }
        } catch (e: Exception) {
            ToolResultPayload(
                ok = false,
                error = "Command execution failed: ${e.message}",
            )
        }
    }

    /**
     * Run a command in a visible IDE terminal (for interactive use).
     */
    fun runInTerminal(command: String, name: String?) {
        // IntelliJ Terminal tool window integration
        // For now, delegates to ProcessBuilder in a visible way
        val workingDir = project.basePath?.let { File(it) } ?: File(System.getProperty("user.dir"))

        ProcessBuilder()
            .command("sh", "-c", command)
            .directory(workingDir)
            .inheritIO()
            .start()
    }
}

File: packages/vscode/src/tools/terminal.ts
import * as vscode from 'vscode';
import type { ToolResultPayload } from '@workflow-extension/shared';

/**
 * Execute terminal commands using VS Code's integrated terminal API.
 *
 * Creates a terminal, sends the command, and captures output.
 * Note: VS Code terminal API does not provide direct output capture,
 * so we use a pseudo-terminal approach for interactive commands,
 * or exec for simple commands with output capture.
 */
export class TerminalOps {
  /**
   * Run a shell command and return its output.
   *
   * Uses child_process.exec via a hidden terminal approach.
   * For long-running commands, uses a visible terminal with notification.
   */
  async runCommand(command: string, cwd?: string): Promise<ToolResultPayload> {
    try {
      const execPromise = new Promise<{ stdout: string; stderr: string }>((resolve, reject) => {
        const cp = require('child_process') as typeof import('child_process');
        cp.exec(
          command,
          {
            cwd: cwd ?? vscode.workspace.workspaceFolders?.[0]?.uri.fsPath,
            maxBuffer: 1024 * 1024, // 1MB
            timeout: 120_000, // 2 minutes
          },
          (err: Error | null, stdout: string, stderr: string) => {
            if (err) {
              reject(err);
            } else {
              resolve({ stdout, stderr });
            }
          },
        );
      });

      const { stdout, stderr } = await execPromise;

      const output = [stdout, stderr].filter(Boolean).join('\n');
      return {
        ok: true,
        content: output || '(command completed with no output)',
      };
    } catch (err: unknown) {
      const execErr = err as { stdout?: string; stderr?: string; message?: string };
      const output = [execErr.stdout, execErr.stderr].filter(Boolean).join('\n');
      return {
        ok: false,
        content: output,
        error: execErr.message ?? String(err),
      };
    }
  }

  /**
   * Run a command in a visible VS Code terminal (for interactive use).
   */
  runInTerminal(command: string, name?: string): void {
    const terminal = vscode.window.createTerminal({
      name: name ?? 'Hermes',
      cwd: vscode.workspace.workspaceFolders?.[0]?.uri.fsPath,
    });
    terminal.show();
    terminal.sendText(command);
  }
}
